### ISSUE TITLE
N_OSO symbols for packages in IssuesTeset.scala are missing in the final binary

### ISSUE THREAD

**Issue by LeeTibbert** [scala-native/scala-native#3458]
maxOS SN build has begun throwing maybe 50 or 100 errors of the form:

```
[info] Linking native code (immix gc, none lto) (427 ms)
[error] warning: (arm64)  could not find object file symbol for symbol __SM33scala.scalanative.issue1359.Main$D15$anonfun$main$1nEPT33scala.scalanative.issue1359.Main$
[error] warning: (arm64)  could not find object file symbol for symbol __SM33scala.scalanative.issue1359.Main$D1fL16java.lang.ObjectEO
```

These started happening Friday2023-08-25, probably sometime after 10:00 UTC-4.
I do not see them on Linux (ubuntu).

scala 2.12.18, java 8
Apple M1,  most up-to-date macOS

git log last commit 
```

commit 54901f5bb0fddf526fbea514de22773b03a5f25c (HEAD -> PR_jl_UnixProcessGen2WaitForExitCode_I3452, upstream/main, origin/main, main)
Author: Wojciech Mazur <wmazur@virtuslab.com>
Date:   Fri Aug 25 21:52:10 2023 +0200

    Better backtrace in classloading for unreachable symbols (#3449)
    
    * Improve reporting unreachable symbols, limit scope of data returned by the linker which would not be used in successful/failing workflow
    * Add minimal tests for missing symbols

```

Generated binary seems to work; well, not quite work yet but it does  get back to the 
bug I am working on.

**Comment by WojciechMazur** [comment]
Actually it's probably due to #3343, I've seen this as well on the M1.

**Comment by tanishiking** [comment]
Interesting, I'd been working on M2 mac and haven't seen the warning 👀 I'll double check is there a problem on machines on my hand (intel mac + M2)

**Comment by tanishiking** [comment]
```sh
❯ cat test.sh
!(sbt "clean;tests3 / Test / nativeLink" | grep -q error)

❯ git bisect start HEAD 4c4006c6534699db3697bc376b65d05f5dab4aab

❯ git bisect run ./test.sh

❯ git bisect log
# bad: [b88b1d4afd52b96d2d33802ea0ad400bf1049dd9] Cleanup stacktraces to not contain intrinsic method calls (#3456)
# good: [4c4006c6534699db3697bc376b65d05f5dab4aab] Fixing Bug in Lowering: Handling Op.Copy for ClassOf Transformation (#3447)
git bisect start 'HEAD' '4c4006c6534699db3697bc376b65d05f5dab4aab'
# bad: [7460645f44bf69fdb0dd708efba8113eec53733a] Build: Update sbt to 1.9.4 (#3451)
git bisect bad 7460645f44bf69fdb0dd708efba8113eec53733a
# bad: [723aef61b9869c182607a4a47f2a7afcdd6ba3aa] Add filename and line number to backtrace on MacOs (#3343)
git bisect bad 723aef61b9869c182607a4a47f2a7afcdd6ba3aa
# good: [1d34e35569da8fd30c7c181cabd2b9f0cd8ab0a2] Allow cross compiling of native tools (#3233)
git bisect good 1d34e35569da8fd30c7c181cabd2b9f0cd8ab0a2
# first bad commit: [723aef61b9869c182607a4a47f2a7afcdd6ba3aa] Add filename and line number to backtrace on MacOs (#3343)
```

And indeed it's caused by https://github.com/scala-native/scala-native/pull/3343 🤔

**Comment by tanishiking** [comment]
- It looks like scala-native runs `dsymutil` against the linked executable in `tests` project.
- And `dsymutil` warns "could not find object file symbol for symbol ..." because the symbol (found from the symtab in the executable) does not have a corresponding debug symbol in any object files. 
- This is because `tests` project does not generate debug metadata since we do not set `withDebugMetadata(true)` in this project.
- **It is strange that we are running `dsymutil` in tests because it should not run if we do not generate debug metadata.**
  - https://github.com/scala-native/scala-native/blob/b88b1d4afd52b96d2d33802ea0ad400bf1049dd9/tools/src/main/scala/scala/scalanative/build/Build.scala#L179-L181C8

**Comment by tanishiking** [comment]
Ah, no we generate debug metadata https://github.com/scala-native/scala-native/blob/b88b1d4afd52b96d2d33802ea0ad400bf1049dd9/project/MyScalaNativePlugin.scala#L68 in `tests`
The problem is still caused by the lack of debug symbol in object files.

**Comment by tanishiking** [comment]
They are defined as global variables in LLVM IR

```
❯ grep -H "issue1359.Main\$D1fL16java.lang.ObjectEO" unit-tests/native/.3/target/scala-3.1.3/**/*.ll
unit-tests/native/.3/target/scala-3.1.3/native-test/scala/scalanative/issue1359.ll:  %_2000006 = bitcast i8* bitcast (i8* (i8*)* @"_SM33scala.scalanative.issue1359.Main$D1fL16java.lang.ObjectEO" to i8*) to i8* (i8*)*
unit-tests/native/.3/target/scala-3.1.3/native-test/scala/scalanative/issue1359.ll:define i8* @"_SM33scala.scalanative.issue1359.Main$D1fL16java.lang.ObjectEO"(i8* %_1) personality i8* bitcast (i32 (...)* @__gxx_personality_v0 to i8*) !dbg !13 {
unit-tests/native/.3/target/scala-3.1.3/native-test/scala/scalanative/issue1359.ll:!13 = distinct !DISubprogram(name: "f", linkageName: "_SM33scala.scalanative.issue1359.Main$D1fL16java.lang.ObjectEO", scope: !4, file: !3, line: 659, type: !7, unit: !4, spFlags: DISPFlagDefinition)
```

and indeed we don't generate debug info for them.

**Comment by tanishiking** [comment]
Nah, it doesn't seem to related to `DIGlobalVariable`, `_SM33scala.scalanative.issue1359.Main$D15main$$anonfun$1L16java.lang.ObjectEPT33scala.scalanative.issue1359.Main$` has it's debug information as `DISubprogram`

```
define dereferenceable_or_null(8) i8* @"_SM33scala.scalanative.issue1359.Main$D15main$$anonfun$1L16java.lang.ObjectEPT33scala.scalanative.issue1359.Main$"(i8* %_1) personality i8* bitcast (i32 (...)* @__gxx_personality_v0 to i8*) !dbg !8 {
...
!8 = distinct !DISubprogram(name: "main$$anonfun$1", linkageName: "_SM33scala.scalanative.issue1359.Main$D15main$$anonfun$1L16java.lang.ObjectEPT33scala.scalanative.issue1359.Main$", scope: !4, file: !3, line: 663, type: !7, unit: !4, spFlags: DISPFlagDefinition)
```

and the object file actually contains the DWARF DIE for that

```
❯ llvm-dwarfdump unit-tests/native/.3/target/scala-3.1.3/native-test/scala/scalanative/issue1359.ll.o
...
0x0000002f:   DW_TAG_subprogram
                DW_AT_low_pc	(0x0000000000000000)
                DW_AT_high_pc	(0x0000000000000034)
                DW_AT_APPLE_omit_frame_ptr	(0x01)
                DW_AT_frame_base	(DW_OP_reg31 WSP)
                DW_AT_MIPS_linkage_name	("_SM33scala.scalanative.issue1359.Main$D15main$$anonfun$1L16java.lang.ObjectEPT33scala.scalanative.issue1359.Main$")
                DW_AT_name	("main$$anonfun$1")
                DW_AT_decl_file	("/Users/tanishiking/ghq/github.com/tanishiking/scala-native/unit-tests/native/src/test/scala/scala/scalanative/IssuesTest.scala")
                DW_AT_decl_line	(663)
                DW_AT_type	(0x0000011b "ptr")
                DW_AT_external	(0x01)
```

**Comment by tanishiking** [comment]
Ok, I have an theory (based on [bugfix #279: Extremely slow dsymutil after Xcode 9.3 update · dkimitsa's RoboVM](https://dkimitsa.github.io/2018/04/04/fix279-extremely-slow-dsymutil/))

## TL;DR
- The Darwin linker distinguishes compilation unit DIEs (debugging information entries) by their `DW_AT_name`, `DW_comp_dir` attribute, and the object files' timestamps.
- Scala Native can generate multiple object files that have the same compilation unit DIE (CU DIE) if a Scala file has multiple top-level packages.
- If the CU DIEs and timestamps are duplicated, the Darwin linker cannot distinguish the DIEs, and one of the duplicates will be ignored.
- As a result, the `N_OSO` symbol (which points to the object file path) is missing in the final binary. Additionally, the `dsymutil` will fail to locate the object file.

## Why `dsymutil` complaints
### Background: Darwin linker
The Darwin linker does not link debugging information into the final binary. Instead, it emits `N_OSO` (object file symbol) type symbol entries in the output binary's symbol table, containing absolute path to the linked-in object files.

e.g.
```sh
❯ dsymutil --symtab unit-tests/native/.3/target/scala-3.1.3/tests-test | grep N_OSO | grep issue1359
[ 57606] 00762e48 66 (N_OSO        ) 00     0001   0000000064f08705 '/Users/tanishiking/ghq/github.com/tanishiking/scala-native/unit-tests/native/.3/target/scala-3.1.3/native-test/scala/scalanative/issue1359.ll.o'
```

Using that reference, 'dsymutil' will link the debug data from the object files into a `dSYM` bundle.

### Lack of `N_OSO` symbol entry
The `dsymutil` complains `could not find object file symbol for symbol ...issue1359.Main$`.
This is because the linked final binary is missing a `N_OSO` symbol entry that points to the object file `unit-tests/native/.3/target/scala-3.1.3/native-test/scala/scalanative/issue1359.ll.o`.

So, why the `dsymutil` could not find the object file `unit-tests/native/.3/target/scala-3.1.3/native-test/scala/scalanative/issue1359.ll.o` even though it exists and contains debug information? - This is because the linked final binary is missing a `N_OSO` symbol entry that points to `issue1359.ll.o`.

Actually, we can't find `issue1359.ll.o` from `dsymutil --symtab unit-tests/native/.3/target/scala-3.1.3/tests-test | grep N_OSO`.

### Why `N_OSO` symbol is missing?
The next question is why is `N_OSO` symbol that points to `issue1359.ll.o` missing from the final binary?

According to the doc comment

> the N_OSO entries contain the absolute path to the file, and the file's timestamp
> However, this doesn't always uniquely identify an object within
> an archive -- an archive file can have multiple entries with the
 > same filename. (This will happen commonly if the original object
 > files started in different directories.) The only way they get
 > distinguished, then, is via the timestamp. But this process is
 > unable to find the correct object file in the archive when there
 > are two files of the same name and timestamp.
https://github.com/llvm/llvm-project/blob/e2cb07c322e85604dc48f9caec52b3570db0e1d8/llvm/lib/Object/ArchiveWriter.cpp#L709-L721

and according to the blog post

> when linking it (Darwin linker) uses DWARF debug information: `AT_name` and `AT_comp_dir` to produce `N_OSO`
> https://dkimitsa.github.io/2018/04/04/fix279-extremely-slow-dsymutil/

While the LLVM documentation mentions the archive, in the case of Scala Native, multiple object files can have the same `DW_AT_name` and `DW_AT_comp_dir` attributes in the CU DIE, even when they are not in an archive.

For example, if a Scala file has multiple package objects, Scala Native will generate multiple object files that have the same CU DIE.

```scala
// Test.scala
package A { ... } // A.ll.o whose compilation unit DIE points to Test.scala
package B { ... } // B.ll.o whose compilation unit DIE points to Test.scala
package C { ... } // same
```

In this case, the Darwin linker distinguishes the DIE entries by their object file timestamps.
However, the `IssuesTest.scala` file has a lot of top-level packages, which means that it generates a lot of object files (`issue1950.ll.o`, `issue2519.ll.o`, and so on) that have the same CU DIE (that points to `IssuesTest.scala`).

If the timestamps of the object files for the packages are the same, the Darwin linker will not be able to distinguish between the DIEs, and one of the duplicates will be ignored, **I guess.**
This means that the `N_OSO` symbol for that object file will not be created.


### Checking
Actually, some object files have the same timestamp with the same CU DIE (`IssuesTest.scala`).

```sh
❯ ls -lT unit-tests/native/.3/target/scala-3.1.3/native-test/scala/**/issue*.ll.o

...
-rw-r--r--  1 tanishiking  staff   3368 Aug 31 21:04:39 2023 unit-tests/native/.3/target/scala-3.1.3/native-test/scala/scalanative/issue1155.ll.o
-rw-r--r--  1 tanishiking  staff   6760 Aug 31 21:04:39 2023 unit-tests/native/.3/target/scala-3.1.3/native-test/scala/scalanative/issue1359.ll.o
...
```




## Solution
### Solution1: Name differently to the compilation unit DIE for packages
The problem is the multiple object files' CU DIE has the same `DW_AT_name` and `DW_AT_comp_dir` + timestamp.
We can avoid this problem if we name differently for the CU DIEs, even those object files are originated from the same Scala file.

For example, maybe we can add a package name to the `DW_AT_name`.

<details>
<summary>Actually, we could get rid of the warnings by appending random number to DW_AT_name</summary>

```diff
diff --git a/tools/src/main/scala/scala/scalanative/codegen/llvm/MetadataCodeGen.scala b/tools/src/main/scala/scala/scalanative/codegen/llvm/MetadataCodeGen.scala
index aa9f2872..7eb4c7fb 100644
--- a/tools/src/main/scala/scala/scalanative/codegen/llvm/MetadataCodeGen.scala
+++ b/tools/src/main/scala/scala/scalanative/codegen/llvm/MetadataCodeGen.scala
@@ -14,6 +14,7 @@ import scala.scalanative.nir.Unmangle
 import scala.scalanative.nir.Global.Member
 import scala.scalanative.nir.Sig.Method
 import scala.scalanative.util.unreachable
+import _root_.java.util.Random
 
 // scalafmt: { maxColumn = 100}
 trait MetadataCodeGen { self: AbstractCodeGen =>
@@ -548,9 +549,10 @@ object MetadataCodeGen {
           .field("language", "DW_LANG_C_plus_plus".const)
     }
 
+    val rand = new scala.util.Random
     implicit lazy val ofDIFile: Specialized[DIFile] = {
       case DIFile(filename, directory) =>
-        _.field("filename", filename)
+        _.field("filename", filename + rand.nextString(5))
           .field("directory", directory)
     }

```

</details>


### Solution2: Just leave it as is, users shouldn't define a lot of packages in one Scala file.

If we don't define multiple packages in one Scala file, there's less chance to generate multiple object files that has the same timestamp + CU DIE.

<details>
<summary>Actually, comment out some packages eliminated the warnings</summary>

```diff
diff --git a/unit-tests/native/src/test/scala/scala/scalanative/IssuesTest.scala b/unit-tests/native/src/test/scala/scala/scalanative/IssuesTest.scala
index 8c3f043f..79b1fafb 100644
--- a/unit-tests/native/src/test/scala/scala/scalanative/IssuesTest.scala
+++ b/unit-tests/native/src/test/scala/scala/scalanative/IssuesTest.scala
@@ -339,9 +339,9 @@ class IssuesTest {
     assertTrue(c.init == "foobar")
   }
 
-  @Test def test_Issue1155(): Unit = {
-    assertTrue(issue1155.C.CLASS.toString.contains("C$CLASS$@"))
-  }
+  // @Test def test_Issue1155(): Unit = {
+  //   assertTrue(issue1155.C.CLASS.toString.contains("C$CLASS$@"))
+  // }
 
   @Test def test_Issue1090(): Unit = {
     val xs = new Array[issue1090.X](20)
@@ -504,22 +504,22 @@ class IssuesTest {
     )
   }
 
-  @Test def test_Issue2520(): Unit = {
-    import issue2520._
-    import issue2520.Kleisli.KleisliOpt
-    implicit def instance: Strong[Function1] = new Strong[Function1] {
-      override def first[A, B, C](f: A => B): Function1[(A, C), (B, C)] = {
-        case (a1, a2) => (f(a1), a2)
-      }
-    }
+  // @Test def test_Issue2520(): Unit = {
+  //   import issue2520._
+  //   import issue2520.Kleisli.KleisliOpt
+  //   implicit def instance: Strong[Function1] = new Strong[Function1] {
+  //     override def first[A, B, C](f: A => B): Function1[(A, C), (B, C)] = {
+  //       case (a1, a2) => (f(a1), a2)
+  //     }
+  //   }
 
-    assertNotNull {
-      val res =
-        implicitly[Strong[KleisliOpt]].law.test[Int, Int, Int, Int](x => x)
-      println(res)
-      res
-    }
-  }
+  //   assertNotNull {
+  //     val res =
+  //       implicitly[Strong[KleisliOpt]].law.test[Int, Int, Int, Int](x => x)
+  //     println(res)
+  //     res
+  //   }
+  // }
 
   @Test def test_Issue2550(): Unit = {
     // The code was failing to link
@@ -638,6 +638,7 @@ package issue1090 {
   }
 }
 
+/*
 package issue1155 {
   trait C {
     def foo = "bar"
@@ -647,6 +648,7 @@ package issue1155 {
     object CLASS extends C
   }
 }
+*/
 
 package issue900 {
   class C(any: Any) {
@@ -689,6 +691,7 @@ package issue2519 {
   object E extends B
 }
 
+/*
 package issue2520 {
   case class Kleisli[F[_], A, B](run: A => F[B])
 
@@ -717,6 +720,7 @@ package issue2520 {
     def law: Law = new Law {}
   }
 }
+*/
 
 package object issue2552 {
   def foo(a: Int, b: Int = 0): Int = 0
```

### Solution3: Slow down clang invocation
If we don't generate multiple object files in the same timestamp (same second on OSX?), we shouldn't have the problem.

**Comment by ekrich** [comment]
Have you looked at what they are doing in RoboVM? <https://github.com/MobiVM/robovm/tree/master/compiler/compiler/src/main/java/org/robovm/compiler/llvm/debug/dwarf>

**Comment by WojciechMazur** [comment]
Adding a suffix to filename does not feel right, as it might lead to some undefined issues later. Similary slowing toolchain is not an option, since it already takes too long time. Leaving it as it it is not a good idea as well, since the ammount of error messages might become overwhelming. 

Becouse the problem arises when multiple compile units (.ll files) point to the some file (nir.Position.file), maybe changing the grouping of defns would be the option? Currently, we group definitions based on their package, which had a good enough balance when it comes to compilation times when compared to emiting N files, where N=available CPUs (best) and emitting an `.ll` file for each class, or rather `Global.Top` (worst performance). 
Maybe, instead of package name, we can group them by the file from the `nir.Position` info attached to each `Defn`? To keep the the good granulity we can try to split them by directory of the file from `nir.Position`. The definitions without nir.Position assigned (generated ones) would be grouped in the one additional file. 
If I understood the problem correctly, it should fix an issue.

**Comment by tanishiking** [comment]
> Have you looked at what they are doing in RoboVM?

Unfortunately, RoboVM doesn't have the same problem with us, they had the same warnings because they are misplacing the `AT_comp_dir` and `AT_name`.

> Maybe, instead of package name, we can group them by the file from the nir.Position info attached to each Defn? To keep the the good granulity we can try to split them by directory of the file from nir.Position. The definitions without nir.Position assigned (generated ones) would be grouped in the one additional file.
If I understood the problem correctly, it should fix an issue.

Yeah, I think that should fix the issue.

---

Before going forward to the @WojciechMazur idea of grouping the `.ll` files by Scala file, I will take a look bit more is it possible for Darwin linker to distinguish the CU DIEs based on another field than `AT_name` and `AT_comp_dir`.
